### DISCUSSION TITLE
Example "weather04.py" is not working as expected

### DISCUSSION THREAD

**Discussion by MikoAL** [Textualize/textual#4253] [Q&A] [answered]
This is the code for the file "weather04.py" from the textual\docs\examples\guide\workers folder

```python
import httpx
from rich.text import Text

from textual import work
from textual.app import App, ComposeResult
from textual.containers import VerticalScroll
from textual.widgets import Input, Static
from textual.worker import Worker


class WeatherApp(App):
    """App to display the current weather."""

    CSS_PATH = "weather.tcss"

    def compose(self) -> ComposeResult:
        yield Input(placeholder="Enter a City")
        with VerticalScroll(id="weather-container"):
            yield Static(id="weather")

    async def on_input_changed(self, message: Input.Changed) -> None:
        """Called when the input changes"""
        self.update_weather(message.value)

    @work(exclusive=True)
    async def update_weather(self, city: str) -> None:
        """Update the weather for the given city."""
        weather_widget = self.query_one("#weather", Static)
        if city:
            # Query the network API
            url = f"https://wttr.in/{city}"
            async with httpx.AsyncClient() as client:
                response = await client.get(url)
                weather = Text.from_ansi(response.text)
                weather_widget.update(weather)
        else:
            # No city, so just blank out the weather
            weather_widget.update("")

    def on_worker_state_changed(self, event: Worker.StateChanged) -> None:
        """Called when the worker state changes."""
        self.log(event)


if __name__ == "__main__":
    app = WeatherApp()
    app.run()
```

## Issue:
The program lags as if there were no async or worker being used.

## Question:
Do I have something configured wrong? Or should I open an issue? orz

**Comment by davep** [comment]
I just gave it a quick spin and it seems to be working fine to me. Perhaps you could start with giving more information about how and where you're running it; perhaps show the output of `textual diagnose` if you have `textual-dev` installed. It might also be useful to get more information on what "lags" actually means so we know what we're supposed to be seeing.

**Reply by MikoAL** [reply]
Let me elaborate using the guide's words
"If you were to run this app, you should see weather information update as you type. But you may find that the input is not as responsive as usual, with a noticeable delay between pressing a key and seeing it echoed in screen. This is because we are making a request to the weather API within a message handler, and the app will not be able to process other messages until the request has completed (which may be anything from a few hundred milliseconds to several seconds later).

To resolve this we can use the [run_worker](https://textual.textualize.io/api/dom_node/#textual.dom.DOMNode.run_worker) method which runs the update_weather coroutine (async def function) in the background. Here's the code:"

Also my textual diagnose:
# Textual Diagnostics

## Versions

| Name    | Value  |
|---------|--------|
| Textual | 0.52.1 |
| Rich    | 13.7.0 |

## Python

| Name           | Value                                          |
|----------------|------------------------------------------------|
| Version        | 3.11.5                                         |
| Implementation | CPython                                        |
| Compiler       | MSC v.1916 64 bit (AMD64)                      |
| Executable     | C:\Users\User\miniconda3\envs\Lucid\python.exe |

## Operating System

| Name    | Value      |
|---------|------------|
| System  | Windows    |
| Release | 10         |
| Version | 10.0.22631 |

## Terminal

| Name                 | Value           |
|----------------------|-----------------|
| Terminal Application | vscode (1.87.0) |
| TERM                 | *Not set*       |
| COLORTERM            | truecolor       |
| FORCE_COLOR          | *Not set*       |
| NO_COLOR             | *Not set*       |

## Rich Console options

| Name           | Value                |
|----------------|----------------------|
| size           | width=219, height=17 |
| legacy_windows | False                |
| min_width      | 1                    |
| max_width      | 219                  |
| is_terminal    | True                 |
| encoding       | utf-8                |
| max_height     | 17                   |
| justify        | None                 |
| overflow       | None                 |
| no_wrap        | False                |
| highlight      | None                 |
| markup         | None                 |
| height         | None                 |

**Reply by davep** [reply, marked as answer]
Sure, I get what the docs say, I was asking how you're seeing and experiencing this lag. For example, here's me (on macOS, not Windows 10, as I don't have Windows 10 to hand) running the above code and it seems to be responding just fine:

https://github.com/Textualize/textual/assets/28237/cbf13852-f579-4797-8f3b-e97ddc97303e

this is in stark contrast to running `weather01.py` (you can see it really struggling when I request cities that the backend obviously hasn't had a chance to cache):

https://github.com/Textualize/textual/assets/28237/98ffe33f-ae0d-420a-aa14-76bbee724740

**Reply by MikoAL** [reply]
Oh! I understand now, there will still be a bit of stuttering even with async or threading, but it is a lot better than nothing.
Thank you for explaining this to me!

**Comment by willmcgugan** [comment]
Might just be the API was slow at the time? It's a free service, and isn't always responsive.

**Reply by MikoAL** [reply]
I'm sorry for my poor wordings, I'll try to be more clear. 

Quote from the guide:
"When an app requests data from the network it is important that it doesn't prevent the user interface from updating. In other words, the requests should be concurrent (happen at the same time) as the UI updates."

The issue I'm encountering is that the UI updates seems to stop when it is calling to the API, which shouldn't happen even if the API is slow.
